File: sdks/python/applad/vectors.py
"""Vectors service."""


class Vectors:
    def __init__(self, client):
        self.client = client

    def create_index(self, index_id: str, dimensions: int, metric: str = None, name: str = None):
        data = {"indexId": index_id, "dimensions": dimensions}
        if metric is not None:
            data["metric"] = metric
        if name is not None:
            data["name"] = name
        return self.client._call("POST", "/vectors/indexes", data)

    def list_indexes(self):
        return self.client._call("GET", "/vectors/indexes")

    def get_index(self, index_id: str):
        return self.client._call("GET", f"/vectors/indexes/{index_id}")

    def delete_index(self, index_id: str):
        return self.client._call("DELETE", f"/vectors/indexes/{index_id}")

    def upsert(self, index_id: str, vectors: list):
        return self.client._call("POST", f"/vectors/indexes/{index_id}/vectors", {"vectors": vectors})

    def query(self, index_id: str, vector: list, top_k: int = None, filter: dict = None):
        data = {"vector": vector}
        if top_k is not None:
            data["topK"] = top_k
        if filter is not None:
            data["filter"] = filter
        return self.client._call("POST", f"/vectors/indexes/{index_id}/query", data)

    def delete_vectors(self, index_id: str, ids: list):
        return self.client._call("POST", f"/vectors/indexes/{index_id}/delete", {"ids": ids})

File: sdks/dart/lib/src/server/vectors.dart
import '../server_client.dart';

class Vectors {
  final ApplAdServer _client;

  Vectors(this._client);

  Future<Map<String, dynamic>> createIndex({
    required String indexId,
    required int dimensions,
    String? metric,
    String? name,
  }) async {
    return _client.post('/v1/vectors/indexes', data: {
      'indexId': indexId,
      'dimensions': dimensions,
      if (metric != null) 'metric': metric,
      if (name != null) 'name': name,
    });
  }

  Future<Map<String, dynamic>> listIndexes() async {
    return _client.get('/v1/vectors/indexes');
  }

  Future<Map<String, dynamic>> getIndex(String indexId) async {
    return _client.get('/v1/vectors/indexes/$indexId');
  }

  Future<Map<String, dynamic>> deleteIndex(String indexId) async {
    return _client.delete('/v1/vectors/indexes/$indexId');
  }

  Future<Map<String, dynamic>> upsert({
    required String indexId,
    required List<Map<String, dynamic>> vectors,
  }) async {
    return _client.post('/v1/vectors/indexes/$indexId/vectors', data: {
      'vectors': vectors,
    });
  }

  Future<Map<String, dynamic>> query({
    required String indexId,
    required List<double> vector,
    int? topK,
    Map<String, dynamic>? filter,
  }) async {
    return _client.post('/v1/vectors/indexes/$indexId/query', data: {
      'vector': vector,
      if (topK != null) 'topK': topK,
      if (filter != null) 'filter': filter,
    });
  }

  Future<Map<String, dynamic>> deleteVectors({
    required String indexId,
    required List<String> ids,
  }) async {
    return _client.post('/v1/vectors/indexes/$indexId/delete', data: {
      'ids': ids,
    });
  }
}
